### PR TITLE
Support for server sent events

### PR DESCRIPTION
**Pull request by tilgovi**
Add server-side support for server sent events, allowing clients to
subscribe to searches.

The implementation relies on Elasticsearch percolator documents and
ephemeral nsq topics. Each client creates an ephemeral topic and a
percolator. The percolator stores the query from the request parameters
and whenever writes occur the annotation document is percolated and
sent to all matching topics.

The TTL feature of Elasticsearch ensures that percolator documents
are cleaned up even in the event of a server fault.

### CODE REVIEW

**Comment by tilgovi** [comment]
Rather than trying to make the streamer sane I found that an implementation from scratch was much easier. I opted for server sent events instead of websockets, but this could be extended to support websockets easily.

**Comment by tilgovi** [comment]
This is all the code that's necessary for the server side of real-time search. All of streamer.py, query-parser.coffee, and stream-filter.coffee could disappear. Most of search-filter.coffee can disappear or be made vastly simpler; it just needs to parse the terms into key/value pairs.

**Comment by judell** [comment]
"All of streamer.py, query-parser.coffee, and stream-filter.coffee could
disappear." !!!

**Comment by seanh** [comment]
This looks _great_

**Comment by seanh** [comment]
Server-sent events are really simple to implement both client- and server-side, nice.

**Comment by BigBlueHat** [comment]
:heart: :exclamation: 

One question:
> The implementation relies on Elasticsearch percolator documents and ephemeral nsq topics.

Makes NSQ a development requirement, yeah? Not a problem. Just clarifying.

**Comment by tilgovi** [comment]
That's not a change from the past.

**Comment by BigBlueHat** [comment]
Oh. My bad. I'd thought NSQ (previously) was only required in production.
Sorry for the noise.

**Comment by tilgovi** [comment]
It's required for the real-time "streamer" feature.

File: h/api/models.py [hunk 1]
```diff
@@ -47,6 +48,7 @@ def __acl__(self):
         'tags': {'type': 'string', 'analyzer': 'uni_normalizer'},
         'text': {'type': 'string', 'analyzer': 'uni_normalizer'},
         'deleted': {'type': 'boolean'},
+        'nipsa': {'type': 'boolean'},
```
Comment (BigBlueHat) [inline h/api/models.py]: This is the only thing in this code that seemed out of place, but I guess this is just exposing the `nipsa` flag on the annotation so UI-side code can Do The Right Thing with it?

**Comment by BigBlueHat** [comment]
Other than that one :question: about the new Annotation model `nipsa` field being in this commit, this gets a hearty :+1: from me. Please :shipit: :exclamation:

**Comment by tilgovi** [comment]
It's not about exposing it to clients. It's about putting it on the mapping
explicitly rather than letting ElasticSearch infer it when the first
NIPSA'd annotation is saved.

Percolation can only work on fields that are mapped and I wanted to be sure
it was mapped so I put it in the mapping.

Whether that gets sent to clients or not is a separate question answered
elsewhere in the code. This is just about search and storage.

**Comment by BigBlueHat** [comment]
@tilgovi much thanks for the explanation. What I should rebase this? It looks good to go, and I believe @seanh :+1:'d it.

**Comment by tilgovi** [comment]
No tests.

**Comment by seanh** [comment]
I think this looks like a great change but I wasn't under the impression that it was finished

**Comment by tilgovi** [comment]
It's pretty much done. All that I want now is to check that there is enough linger time on the channel to allow reconnects without interrupting the stream. That's important since most browsers are terrible at streaming xhr and leak memory if you don't periodically reconnect when emulating EventSource.

**Comment by seanh** [comment]
Doesn't this also need a matching client-side part?

**Comment by tilgovi** [comment]
It can be done in two parts. This doesn't remove anything the current
client needs.

**Comment by nickstenning** [comment]
Closing in favour of #2511.

**Pull request closed without merging**